### PR TITLE
Adhoc NPM Release GitHub Action

### PR DESCRIPTION
**Pull request by GregBrimble**
This removes the previous `pages-ci.yml` GitHub Action and instead replaces it with a more generic `create-pullrequest-prerelease.yml` (and `delete-pullrequest-prerelease.yml`).

This new workflow automatically builds and publishes a wrangler tgz for every commit of a pull request. It is pushed to KV where anyone can `npm install https://link/wrangler.tgz`. These prereleases are automatically cleaned up once the PR is closed.

A sticky comment appears on the PR with the latest link:

image

The link itself is `https://host/download/<pr number>-<head branch>/<sha>/wrangler.tgz`. The latest version is also aliased at `https://host/download/<pr number>-<head branch>/wrangler.tgz`

This means people can develop and quickly test out new features and bug-fixes without each setting up a workflow like we had for the `pages` tag.

### CODE REVIEW

**Comment by petebacondarwin** [comment]
I think that a better approach, rather than actually releasing to npm is to use Github artifacts.
We can upload build outputs from GH actions, which are then available for up to 90 days.
On the Angular project we do this for every PR: the CI builds the npm packages and uploads them (using CircleCI) so that anyone can go and grab them via the URL. See https://app.circleci.com/pipelines/github/angular/angular/41688/workflows/41931026-96fd-423c-84bb-f9d966f3b53a/jobs/1108916/artifacts

It is easy to use these files in package.json - you just give the URL of the `.tgz` file.

That saves us from polluting the npm repository with hundreds of releases - and actually makes it even easier for people to access builds for testing locally.

The Github Action https://github.com/actions/upload-artifact should make this straightforward to setup. Here is more info about this https://docs.github.com/en/actions/advanced-guides/storing-workflow-data-as-artifacts

**Comment by GregBrimble** [comment]
Yeah, great idea. I’ll look into it this weekend.

**Comment by GregBrimble** [comment]
This might be a dealbreaker: https://github.com/actions/upload-artifact/issues/109

**Comment by GregBrimble** [comment]
Maybe we upload to KV instead and host a little Worker service?

**Comment by petebacondarwin** [comment]
> This might be a dealbreaker: [actions/upload-artifact#109](https://github.com/actions/upload-artifact/issues/109)

😢 that is a really unfortunate setup.

**Comment by petebacondarwin** [comment]
> Maybe we upload to KV instead and host a little Worker service?

That's a good idea! A good dog-food of Workers and seems like a good fit for this use-case.

**Comment by petebacondarwin** [comment]
Maybe we could publish our own GH action, which other people could use to push to their own KV store, as a way to drive more use of Workers?

File: .github/workflows/adhoc-prerelease.yml
Comment (threepointone) [inline .github/workflows/adhoc-prerelease.yml:8]: maybe this should mandatory? We'll never want to do a production release with this, yes?
```suggestion
        required: true
```

Comment (GregBrimble) [reply .github/workflows/adhoc-prerelease.yml:8]: Good point!

Comment (GregBrimble) [reply .github/workflows/adhoc-prerelease.yml]: I did think about some of this, but anyone in the npm org could just publish locally if they wanted. So I figure we trust everyone who is a maintainer of this repo?

Yeah, the primary reason this arose was that I wanted to test more than one thing, so kind of out-grew the Pages workflow. Could definitely keep it if you're happy to start amassing several branches/tags in that though.

File: .github/workflows/adhoc-prerelease.yml
Comment (threepointone) [inline .github/workflows/adhoc-prerelease.yml:55]: Let's just remove this since the tag is now mandatory?
```suggestion
```

File: .github/workflows/adhoc-prerelease.yml
Comment (threepointone) [reply .github/workflows/adhoc-prerelease.yml]: Can this workflow only be run by maintainers? (I actually dunno). That could work. But we should still fail the flow if the input is `latest`, `beta` or `alpha` imo. 

Let's keep the Pages workflow for now and delete once we don't use it for a while, fair? Which is to say, once folks stop making commits there (or is that just you rn?)

File: .github/workflows/adhoc-prerelease.yml
Comment (petebacondarwin) [inline .github/workflows/adhoc-prerelease.yml:8]: I guess that `required: true` allows the empty string?
Otherwise the `if:` further down would have no effect...

**Review by petebacondarwin — APPROVED** [review]
Fine to go with the for the time-being.
But in the long run, I think it would be good to look into uploading PR distributions to some KV store. (Apart from anything, it is a good use of Workers+KV).

**Comment by GregBrimble** [comment]
I'm about 60% done with a KV store. Gimme another day to get that going :)

File: packages/wrangler/package.json
Comment (petebacondarwin) [inline packages/wrangler/package.json:98]: I wonder if we should just move that tsconfig.json here?

Comment (GregBrimble) [reply packages/wrangler/package.json:98]: Could do. Thought it might be re-used with `create-workers-app` though?

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Thanks for putting this together @GregBrimble - it is awesome. I am excited to put it to work.

But first, I think we need to invert things a little bit to make it more secure from malicious attack.

- Currently, as it is implemented, the POST and DELETE APIs are completely open, right?
- That means that anyone could bombard us with massive files and overload our KV store (and Workers).
- Further they could go and in delete our legitimate uploaded artifacts, right?

I think a safer way to do this is for the GH Action to be a kind of webhook to the Worker that just notifies it that something might have changed. Then the Worker will go and interrogate the GH API to find out the status of the PR. If it finds it is closed then it will delete all the artifacts. If it finds there is a commit that is not already in the KV store then it will create a new artifact.

This way, even if someone hit our Worker maliciously, the worst that happens is that it just noops over and over again.

What do you think?

**Comment by GregBrimble** [comment]
Re. security, I'm planning to put Access in front of the upload and delete endpoints. There's a `env.TODO_SECRET_TOKEN` which I'll replace with an Access service token as soon as we know where we're putting this long-term.

I think once this is secure, it'll be easier to have the Worker be a simple CRUD API sort of thing for blob storage and have the logic be in the uploader script rather than calling the GitHub API.

File: packages/prerelease-registry/tsconfig.json
Comment (GregBrimble) [reply packages/prerelease-registry/tsconfig.json:1]: Yeah, I like that last suggestion. I'll break it into three: global, wrangler and prerelease-registry.

File: packages/prerelease-registry/wrangler.toml
Comment (GregBrimble) [reply packages/prerelease-registry/wrangler.toml:3]: Does your team have a shared account? I know the Cloudflare Docs own developers.cloudflare.com. That's maybe a decent second option? (Not sure if they own `developers.workers.dev`?)

File: packages/prerelease-registry/wrangler.toml
Comment (GregBrimble) [reply packages/prerelease-registry/wrangler.toml:5]: Absolutely would be nice if we get CI for it. I want to get this off my personal account before merging, but I reckon we can do CI for the prerelease-registry as a second piece of work.

File: packages/prerelease-registry/functions/delete/[tag].ts
Comment (GregBrimble) [reply packages/prerelease-registry/functions/delete/[tag].ts:16]: 1 per second per key.

File: .gitignore
Comment (GregBrimble) [reply .gitignore:178]: I was getting annoyed with all the other things we were missing. npm debug logs, packed tgz, etc.

gitignore.io provides decent defaults for projects based on the configuration they have set up. In this case, I generated one for Node.JS and macOS.

Happy to go back to a manually curated one if there's any reason to, but this just means we can focus on the project-specific stuff at the end of the file.

File: .github/workflows/create-pullrequest-prerelease.yml
Comment (GregBrimble) [reply .github/workflows/create-pullrequest-prerelease.yml:46]: Thought about that, but thought y'all would push back. More than happy to change it :)

File: packages/prerelease-registry/package.json
Comment (GregBrimble) [reply packages/prerelease-registry/package.json:6]: I can simplify it a bit, actually: that final `functions` isn't necessary. And I guess the `npx` bit isn't necessary, but I quite like that signal that it's a node thing that we're running (vs. something else native).

File: packages/prerelease-registry/wrangler.toml
Comment (petebacondarwin) [reply packages/prerelease-registry/wrangler.toml:3]: 🤔

File: packages/prerelease-registry/package.json
Comment (petebacondarwin) [reply packages/prerelease-registry/package.json]: Don't worry about this case. It was more of a concern for people actually using `wrangler pages` in real world scenarios.

File: packages/prerelease-registry/functions/delete/[tag].ts
Comment (petebacondarwin) [reply packages/prerelease-registry/functions/delete/[tag].ts:16]: But we can hit the KV store with 100s of requests to delete independent keys all at the same time?

**Comment by petebacondarwin** [comment]
> Re. security, I'm planning to put Access in front of the upload and delete endpoints. There's a `env.TODO_SECRET_TOKEN` which I'll replace with an Access service token as soon as we know where we're putting this long-term.
> 
> I think once this is secure, it'll be easier to have the Worker be a simple CRUD API sort of thing for blob storage and have the logic be in the uploader script rather than calling the GitHub API.

I don't think that would be enough. Anyone could create a PR that does bad stuff to that Worker since the PR will be run by our GH Action and so get the auth token automatically.

File: packages/prerelease-registry/functions/delete/[tag].ts
Comment (GregBrimble) [reply packages/prerelease-registry/functions/delete/[tag].ts:16]: Should be fine :)

https://developers.cloudflare.com/workers/platform/limits#kv-limits

We can make 1000 KV operations per request invocation, and there's no limit on concurrency other than 1/second/key.

File: .github/workflows/delete-pullrequest-prerelease.yml
Comment (JacobMGEvans) [reply .github/workflows/delete-pullrequest-prerelease.yml:18]: I see what the token is for further down, ignore my comment lol

**Comment by GregBrimble** [comment]
I think that's now everything other than the open security question. Brain has gone to mush now, so I'll think it over tomorrow and report back.

**Comment by GregBrimble** [comment]
> I don't think that would be enough. Anyone could create a PR that does bad stuff to that Worker since the PR will be run by our GH Action and so get the auth token automatically.

Yeah, I think you're right. But, ultimately, it's the Action that will do the building and packing, so we'll have to have to accept whatever it gives us anyway. So, I propose that we change the registry to be write-only and remove the omit-version aliasing.

1. Action is invoked.
1. Action builds and packs the packages.
1. Action posts to the Worker (unauthenticated) the packages, with PR number, head branch ref and sha.
1. Optionally, the Worker verifies that the request came from an Action. This could be done by also passing the GitHub token with the packages in the previous step, and then the Worker making an authenticated request to GitHub. The token will only be alive as long as the Action is running. We could go one step even further and check that the PR has a currently running action.
1. Worker stores the file and makes it available at a unique URL. The URL is added to a blocklist so it can never be used again.
1. We automatically remove the file after 1 month.

This isn't 100% foolproof. If we implement the full GitHub verification, then someone could still snipe any currently running PR. If we implement the weak GitHub verification, then someone could releases to any PR. And if we implement no GitHub verification, the someone could add to any PR without it having to run on Actions.

I suggest the middle option. It's relatively easy to implement and would be pretty obvious for us to spot anyone malicious. Worst they could do would be to add another package, which won't pose a risk to anyone with it already installed. It also won't update the sticky message on that other PR. Plus, we'll now have their GitHub username which we could block.

**Comment by GregBrimble** [comment]
Okay, new plan. We're going to try hosting things as GitHub Actions artifacts, but put a Worker in front which unzips on the fly. Implementing that today.

**Comment by GregBrimble** [comment]
There's no sane way to look at the GitHub Actions workflow runs associated with a PR, so we unfortunately lose the aliasing. That means every run will have a different URL and there isn't one single always-pointing-at-latest for a PR. No biggie.

The GitHub Action now just builds, packs and uploads the result as a GitHub Action artifact. The sticky comment URL can be calculated ahead-of-time by knowing the run ID, so it just gets updated in isolation.

The Worker is entirely stateless and simply finds the artifacts for a given workflow run, matches by name, unzips and returns the response. We cache the result for performance.

The Worker has a Personal Access Token (mine). We need to authenticate to download artifacts. We can add a proper GitHub App in the future if we care enough, but I think for now, this is good enough.

**Review by petebacondarwin — APPROVED** [review]
This is marvellous. Approving with a bunch of questions and suggestions.
Happy to merge this and iterate, given that it is only for development purposes.
I do think that we should move away from using your personal user info ASAP.

File: packages/prerelease-registry/functions/runs/[[path]].ts
Comment (GregBrimble) [reply packages/prerelease-registry/functions/runs/[[path]].ts:91]: `waitUntil` is not the same as `await`. The former will queue it up as work to perform after the request/response phase is complete. The latter would block.

Without the `waitUntil`, the Workers runtime could kill the promise after the request/response phase is complete.

See more: https://developers.cloudflare.com/workers/runtime-apis/fetch-event#waituntil

File: packages/prerelease-registry/functions/runs/[[path]].ts
Comment (GregBrimble) [reply packages/prerelease-registry/functions/runs/[[path]].ts:20]: Yup, exactly. [Our docs](https://developers.cloudflare.com/workers/runtime-apis/cache#accessing-cache) and [MDN's](https://developer.mozilla.org/en-US/docs/Web/API/Cache).

File: packages/prerelease-registry/functions/runs/[[path]].ts
Comment (GregBrimble) [reply packages/prerelease-registry/functions/runs/[[path]].ts:41]: You're right: that final layer isn't needed. I'll remove it.

But in general to properly clone a Request, you do need to `new Request(request.clone())`. The `.clone()` clones the body, and the `new Request()` makes the headers mutable. Rather annoying, but hey ho.

**Pull request merged**
